lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -674,6 +677,101 @@ postShelleyWallet ctx body = do
     wid = WalletId $ digest $ publicKey rootXPrv
     wName = getApiT (body ^. #name)
 
+toDelegationMsg
+    :: W.EpochNo
+    -> W.SlotParameters
+    -> [W.DelegationDiscovered]
+    -> ApiWalletDelegation
+toDelegationMsg currentEpochNo sp dlgs = case dlgs of
+    [] ->
+        ApiWalletDelegation
+        (ApiDelegationStatus $ ApiT W.NotDelegating)
+        Nothing
+    [(W.DelegationDiscovered (W.SlotId epochN _) poolIdMaybe)] ->
+        case poolIdMaybe of
+            Nothing ->
+                error "there should not be quiting delegation when \
+                      \no prior joining delegation is present"
+            Just poolId ->
+                if currentEpochNo < epochN + 2 then
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT W.NotDelegating)
+                    (Just $
+                        ApiWalletDelegationNext
+                        (ApiDelegationStatus $ ApiT $
+                            W.Delegating (ApiT poolId))
+                        (ApiEpochInfo
+                            (ApiT $ epochN + 2)
+                            (W.epochStartTime sp (epochN + 2)))
+                    )
+                else
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $
+                        ApiT $ W.Delegating (ApiT poolId))
+                    Nothing
+    [ (W.DelegationDiscovered (W.SlotId epochN1 _) poolIdMaybe1)
+        , (W.DelegationDiscovered (W.SlotId epochN2 _) poolIdMaybe2)] ->
+        if epochN2 > epochN1 then
+            error "delegation discovered later cannot have epoch number \
+                  \older than the older discovered delegation"
+        else case (poolIdMaybe1, poolIdMaybe2) of
+            (Nothing, Nothing) ->
+                error "there should not be two quit delegation in row"
```

Comment:
> If we end up here something is seriously wrong

I should clarify: I'm particularly worried because we have had bugs with the chain follower.

Reply:
yeah, I knew you would make a complain here :-) Frankly speaking, every place where is error emitted means we have really huge problem with either DB or how it should work (eg. we cannot quit twice, etc). I need to think more about it frankly speaking :thinking:

Reply:
If we don't like adding proper error handling for "extreme cases", maybe we could `throw` a custom exception and then catch and render as `err501 { message = messageFromException }` somehow.

I.e. 

```haskell
throwInternalError "there should not be two quit delegation in row"
-- instead of 
error "there should not be two quit delegation in row"
```

But that might be a controversial solution.

For listing stake pools, we just have proper error cases for the supposedly impossible. https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Api/Server.hs#L1985-L1991

Reply:
I see. I will add then the error for "the supposedly impossible" also.

Reply:
done in https://github.com/input-output-hk/cardano-wallet/pull/1322/commits/9256bc9d417b474de22dd3315369e4327411dc3b